# CoronaWhy Protein Explorer

## Installation

The entire `explorer` folder can be copied into a Django framework.

### Add to installed apps

Add the explorer to the installed apps list in your `settings.py`:

```
INSTALLED_APPS = (
    ...
    'explorer.apps.ExplorerConfig',
    ...
)
```

### Create URL

Add a URL for the explorer to your `urls.py` using `django.conf.urls.url`:

```
urlpatterns = [
    ...
    url(r'^explorer/', include('explorer.urls')),
    ...
];
```

Or using `django.urls.path`:

```
urlpatterns = [
    ...
    path('explorer/', include('explorer.urls')),
    ...
];
```

### Database Migration (initialization)

Make database migrations with:

```
python manager.py makemigrations explorer
python manager.py migrate explorer
```

### Database population

Populate the database using files from the `explorer/data/imports` folder as follows:

```
# import taxa
python manage.py import_taxa explorer/data/imports/taxonomy_parsed.csv

# import proteins
python manage.py import_protein explorer/data/imports/initial_protein.csv

# import alignments
python manage.py import_alignment explorer/data/imports/initial_alignment.csv

# import sequence records
python manage.py import_sequencerecord explorer/data/imports/spike_sequence_records.csv

# import sequence alignments
python manage.py import_sequence explorer/data/imports/spike_sequence_alignment_20200505.csv

# import epitope alignments
python manage.py import_epitope explorer/data/imports/aligned_epitopes.csv

# import epitope experiments
python manage.py import_epitopeexperiment explorer/data/imports/epitopes_experiments.csv

# import position nomenclature
python manage.py import_nomenclature explorer/data/imports/HKU-N2_nomenclature.csv

# import structures
python manage.py import_structure explorer/data/structure/imports/structureChains.csv

# import structure sequences
python manage.py import_structuresequence explorer/data/structure/imports/structureSequences.csv

# import structure residues and atoms
python manage.py import_structureatom explorer/data/structure/imports/structureAtoms.csv
```

## Further Reading

### Data Gathering and Processing
The data used to populate are generated by scripts in the `explorer/data/` folder.

| Folder | Description |
|--------|-------------|
| `explorer/data/cleaning/` | Manual data cleanup files. |
| `explorer/data/epitope/`  | Epitope data and documentation. |
| `explorer/data/imports/`  | Final data files to import. |
| `explorer/data/protein/`  | Protein sequence data and documentation. |
| `explorer/data/structure/`| Crystal structure data and documentation. |
| `explorer/data/taxon/`    | Taxnomy data and documentation. |

### Manager Commands
The explorer has custom manager commands that can be used to import data etc... They're executed as:

```
python manager.py [COMMAND] [ARGUMENTS]
```

Detailed documentation on each command can be found in the `explorer/management/commands/` folder. The commands are summarized here:

| Command | Description |
|---------|-------------|
| `align_epitopes`              | Align raw epitope sequences with database sequences. |
| `import_alignment`            | Import alignment records. |
| `import_epitope`              | Import aligned epitope sequences. |
| `import_epitopeexperiment`    | Import epitope experiments. |
| `import_nomenclature`         | Import alignment position nomenclatures. |
| `import_protein`              | Import protein records. |
| `import_sequence`             | Import protein sequence alignments. |
| `import_sequencerecord`       | Import protein sequence records. |
| `import_structure`            | Import crystal structure records. |
| `import_structureatom`        | Import atoms and residues of a crystal structure. |
| `import_structuresequence`    | Import crystal structure sequence alignments. |













fin.
