- [Quick Reference](#quick-reference)
- [Protein Sequence Records and Alignments](#protein-sequence-records-and-alignments)
  * [Sequence Records](#sequence-records)
  * [Sequence](#sequence)
  * [Nomenclature](#nomenclature)
- [Immunological Epitopes and Experiments](#immunological-epitopes-and-experiments)
  * [Epitope Experiment Classes](#epitope-experiment-classes)
  * [Epitope Experiments](#epitope-experiments)
  * [Epitope Sequences](#epitope-sequences)
- [Crystal Structures](#crystal-structures)
  * [Crystal Structure Names and Chains](#crystal-structure-names-and-chains)
  * [Crystal Structure Sequence Alignments](#crystal-structure-sequence-alignments)
  * [Crystal Structure Residue Info and Coordinates](#crystal-structure-residue-info-and-coordinates)

## Quick Reference

| Type | URL | Description |
|------|-----|-------------|
| [Sequence Records](#sequence-records) | `explorer/sequencerecord` | Get filtered sequence record metadata by `mesh_id` and additional criteria. |
| [Sequence](#sequence) | `explorer/sequences` | Get sequence alignments by `mesh_id`, `alignment` and `accession`. |
| [Nomenclature](#nomenclature) | `explorer/nomenclature` | Generate sequence alignment nomenclature by `mesh_id`, `alignment` and `accession`. |
| [Epitope Experiment Classes](#epitope-experiment-classes) | `explorer/epitopeexperimentclases` | Get distinct values for epitope experiment data to build select lists, using `mesh_id` and `alignment`.  |
| [Epitope Experiments](#epitope-experiments) | `explorer/epitopeexperimentsfilter` | Get filtered epitope experiment data by `mesh_id`, `alignment`, and additional criteria. |
| [Epitope Sequences](#epitope-sequences) | `exporer/epitopesequence` | Get epitope sequence alignments by `iedb_id`, `mesh_id` and `alignment`. |
| [Crystal Structure Names and Chains](#crystal-structure-names-and-chains) | `explorer/structurechains` | Get list of crystal structure PDBs and chain IDs. by `mesh_id`. |
| [Crystal Structure Sequence Alignments](#crystal-structure-sequence-alignments) | `explorer/structuresequence` | Get crystal structure chain sequence alignments by `mesh_id`, `alignment` and `[PDB].[CHAIN]`. |
| [Crystal Structure Residue Info and Coordinates](#crystal-structure-residue-info-and-coordinates) | `explorer/structureresidueatoms` | Get crystal structure residue atom coordinates and info by `mesh_id`, `alignment` and `[PDB].[CHAIN]`. |

## Protein Sequence Records and Alignments

The API provides a filterable search of sequence records through the `sequencerecord` URL. Accession IDs and alignment names are then used to access the sequence alignments themselves through the `sequences` URL. Position nomenclatures for a given reference sequence (via. Accession) and alignment can also be generated by the `nomenclature` URL.

**Note** that all accession numbers include versions (e.g. XXXXX.# format).

### Sequence Records

#### Usage
`explorer/sequencerecord?mesh_id=[MESH ID]&[OPTIONAL CRITERIA]`

**Example**: http://localhost:8000/explorer/sequencerecords?mesh_id=D064370&host=Homo%20sapiens&country=USA%2CChina

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
|mesh_id            | Y | NCBI MeSH ID of the protein. |
|accession          | N | Comma-separated list of protein accession number(s) with version. |
|organism           | N | Comma-separated list of organism name(s). |
|country            | N | Comma-separated list of country name(s). |
|host               | N | Comma-separated list of host name(s). |
|isolation_source   | N | Comma-separated list of isolate tissue source(s). |
|isolate            | N | Comma-separated list of isolate name(s). |
|taxon_id           | N | Comma-separated list of NCBI taxon ID(s). |

**Note:** All arguments except for mesh_id can be comma-separated lists and will be filtered as "in".

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| taxon_id     | int    | GenBank taxon ID that this sequence record is for. |
| accession    | str    | NCBI Protein accession number (With version). |
| organism     | str    | Organism name. |
| collection_date | str | Sequence collection date. |
| country      | str    | Country, country:region the sequence was obtained from. |
| host         | str    | Name of organism host. |
| isolation_source | str | Name of tissue sequence was isolated from. |
| isolate | str | Name of isolate. |
| coded_by | str | NCBI nucleotide accession and region that codes this protein sequence. |

**Example**:
```
[
    {
        "accession": "QEG03814.1",
        "organism": "Human coronavirus OC43",
        "collection_date": "2019-01-01 00:00:00+00:00",
        "country": "USA",
        "host": "Homo sapiens",
        "isolation_source": null,
        "isolate": "HCoV_OC43/Seattle/USA/SC0776/2019",
        "taxon_id": "31631"
    }, {
        "accession": "QEG03803.1",
        "organism": "Human coronavirus OC43",
        "collection_date": "2019-01-01 00:00:00+00:00",
        "country": "USA",
        "host": "Homo sapiens",
        "isolation_source": null,
        "isolate": "HCoV_OC43/Seattle/USA/SC9428/2018",
        "taxon_id": "31631"
    }, {
        "accession": "ARB66071.1",
        "organism": "Human coronavirus OC43",
        "collection_date": "None",
        "country": "China",
        "host": "Homo sapiens",
        "isolation_source": "nasopharyngeal swab",
        "isolate": "guangzhou/2013/15",
        "taxon_id": "31631"
    },
    ...
]
```

### Sequence

#### Usage
`explorer/sequences?mesh_id=[MESH ID]&alignment=[ALIGNMENT NAME]&accession=[COMMA-SEPARATED ACCESSION NUMBERS]`

**Example**: http://localhost:8000/explorer/sequences?mesh_id=D064370&alignment=20200505&accession=P15777.1%2CABP38295.1%2CABP38267.1%2CP25190.1%2CP25191.1%2CQ9QAQ8.1%2C

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |
| accession | Y | Comma separated list of NCBI Protein accession numbers (including version) |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| accesion  | str | Accession number with version. |
| offset | int | offset of the sequence in alignment. |
| seq | str | aligned sequence without leading gaps. |

**Example**:
```
[{
    "accession": "ABP38295.1",
    "seq": "MFLILLISLPTAFAV---------IGDLKCTT-----LSINDVD-TGV---PSISTDTVDVTNGLGTYYVLDRVYLNTTLLLNGYYPTSGSTYRNMALKGTLLLSTLWF---------------KPPFLSDFTNGIFAKVKNTKVI-----------KDGVMYSEFPAITIGSTFVNTSYSVVV---QPHTTIL----GNKLQG-------FLEISVCQ---------YTMCEYPNTICNPNLG-NQRVELWHWD-TG-VVSCLYKR---NFTYDVNA----DY-----LYFHFYQEGGTFYAYFTDTGVVTK--------------FLF---NVYLGTVLSHYYVM-PLTCNS-------------ALTLEYWVTPLTSKQYLLAFNQDGVIFNAVDCKSDFMSEIKCKTLSIAPSTGVYELNGYTVQPIADVYRRIPNLPDCN--IEAWL--NDKSVPSPLNWERKTFSNCNFNMSSLMSF-IQADSFTCNNIDAAKIYGMCFSSITIDKFAIPNGRKVDLQLGNLGYLQSFNYRIDTTATSCQLYYNLPAANVSVSRFNPSTWNRRFGFTEQSVFKPQPAGVFTDHDVVYAQHCFKAPTNFCPCKLDGSLCVGNGPGIDAGYKTSGIGTCPAGTN-----YLTCHN----------------AAQCDCLCTPDPITSKATGPY---KCPQTKYLVGIGEHCSGLAIKSDYCGGNP-------CSCQPQ------------AFLGWSVD-SCLQGDRCNIFANFILHDVNSGTTCSTDLQ---KSNTDIILGVCVNYDLYGITGQGIFVEVNATYYNSWQNLLYDSNGNLYGFRDYLTNRTFMIRSCYSGRVSAAFHA--NSSEPALLFRNIKCNYVFNNTL---SRQL---------QPINY-FDSYLGCVVNADNSTSIVVQ--TCDLTVGNGYCVDYST-----KRRSRRSITTGYRFTNFEPFTVNSVNDSLEPVGGLYEIQIPSEFTIGNMEEFIQTSSPKVTIDCSAFVCGDYAACKSQLVEYGSFCDNINAILTEVNELLDTTQLQVANSLMNGVTLSTKLKDGVNFNVDDINFSPVLGCLGSDCNKV-------SSRSAIEDLLFSKVKLSDVGFVEAYNNCTG---GAEIRDLICVQSYNGIKVLPPLLSENQISGYTLAATSASLFPPWSA----AAGVPFYLNVQYRINGIGVTMDVLSQNQKLIANAFNNALGAIQEGFDATNSALVKIQAVVNANAEALNNLLQQLSNRFGAISSSLQEILSRLDALEAQAQIDRLINGRLTALNAYVSQQLSDSTLVKFSAAQAMEKVNECVKSQSSRINFCGNGNHIISLVQNAPYGLYFIHFSYVPTKYVTAKVSPGLCIAG--GRGIAPKSGYFVNVNN----------SWMFTGSGYYYPEPIT-GNNVVVMSTCAVNYTKAPDVMLNIST--PNLPDFKEELDQWFKNQTSVAPDLSL----D--YINVTFLDLQDEMNRLQEAIKVLNQSYINLKDIGTYEYYVKWPWYVWLLIGFAGVAMLVLLFFICCCTGCGTSCFKK--CGGCCDDYTGHQELVIK---TSHED",
    "offset": 7
}, {
    "accession": "ABP38267.1",
    "seq": "MFLILLISLPTAFAV---------IGDLKCTT-----VSINDVD-TGV---PSISTDTVDVTNGLGTYYVLDRVYLNTTLLLNGYYPTSGSTYRNMALKGTLLLSTLWF---------------KPPFLSDFTNGIFAKVKNTKVV-----------KDGVMYSEFPAITIGSTFVNTSYSVVV---QPHTTIL----GNKLQG-------FLEISVCQ---------YTMCEYPNTICNPNLG-NQRVELWHWD-TG-VVSCLYKR---NFTYDVNA----DY-----LYFHFYQEGGTFYAYFTDTGVVTK--------------FLF---NVYLGTVLSHYYVM-PLTCNS-------------ALTLEYWVTPLTSKQYLLAFNQDGVIFNAVDCKSDFMSEIKCKTLSIAPSTGVYELNGYTVQPIADVYRRIPNLPDCN--IEAWL--NDKSVPSPLNWERKTFSNCNFNMSSLMSF-IQADSFTCNNIDAAKIYGMCFSSITIDKFAIPNGRKVDLQLGNLGYLQSFNYRIDTTATSCQLYYNLPAANVSVSRFNPSTWNRRFGFTEQSVFKPQPAGVFTDHDVVYAQHCFKAPTNFCPCKLDGSLCVGSGSGIDAGYKNTGIGTCPAGTN-----YLTCHN----------------AAQCNCLCTPDPITSKATGPY---KCPQTKYLVGIGEHCSGLAIKSDHCGGNP-------CSCQPQ------------AFLGWSVD-SCLQGDRCNIFANFILHDVNSGTTCSTDLQ---KSNTDIILGVCVNYDLYGITGQGIFVEVNATYYNSWQNLLYDSNGNLYGFRDYLTNRTFMIRSCYSGRVSAAFHA--NSSEPALLFRNIKCNYVFNNTL---SRQL---------QPINY-FDSYLGCVVNADNSTSSVVQ--TCDLTVGSGYCVDYST-----KRRSRRSITTGYRFTNFEPFTVNSVNDSLEPVGGLYEIQIPSEFTIGNMEEFIQTSSPKVTIDCSAFVCGDYAACKSQLVEYGSFCDNINAILTEVNELLDTTQLQVANSLMNGVTLSTKLKDGVNFNVDDINFSPVLGCLGSDCNKV-------SSRSAIEDLLFSKVKLSDVGFVEAYNNCTG---GAEIRDLICVQSYNGIKVLPPLLSENQISGYTLAATSASLFPPWSA----AAGVPFYLNVQYRINGIGVTMDVLSQNQKLIANAFNNALGAIQEGFDATNSALVKIQAVVNANAEALNNLLQQLSNRFGAISSSLQEILSRLDALEAQAQIDRLINGRLTALNAYVSQQLSDSTLVKFSAAQAMEKVNECVKSQSSRINFCGNGNHIISLVQNAPYGLYFIHFSYVPTKYVTAKVSPGLCIAG--GRGIAPKSGYFVNVNN----------TWMFTGSGYYYPEPIT-GNNVVVMSTCAVNYTKAPDVMLNIST--PNLPDFKEELDQWFKNQTSVAPDLSL----D--YINVTFLDLQDEMNRLQEAIKVLNQSYINLKDIGTYEYYVKWPWYVWLLIGFAGVAMLVLLFFICCCTGCGTSCFKK--CGGCCDDYTGHQELVIK---TSHED",
    "offset": 7
}]
```

### Nomenclature

#### Usage
`explorer/nomenclature?mesh_id=[MESH ID]&alignment=[ALIGNMENT NAME]&accession=[ACCESSION]`

**Example**: http://localhost:8000/explorer/nomenclature?mesh_id=D064370&alignment=20200505&accession=P25193.2

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |
| accession | Y | NCBI Protein accession number (including version) of the nomenclature reference sequence. |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| [0]       | int | Major position. |
| [1]       | int | Minor position. |

**Note**: the index of the outer array is the index of each character in the reference sequence alignment itself.

**Example**:
```
[
    [0, 0],     # major, minor position for first character in aligned string
    [1, 0],     # major, minor position for second character in aligned string
    [2, 0],     # ...
    [3, 0],
    [4, 0],
    [5, 0],
    [6, 0],
    [7, 0],
    [8, 0],
    [9, 0],
    [10, 0],
    [11, 0],
    [12, 0],
    [13, 0],
    [14, 0],
    [15, 0],
    [15, 1],    # first insert at position 15, sequence is a gap (-), minor=1
    [15, 2],    # second insert at position 15, sequence is a gap (-), minor=2
    [15, 3],    # third insert at position 15, sequence is a gap (-), minor=3
    [15, 4],
    [15, 5],
    [15, 6],
    [15, 7],
    [15, 8],
    [15, 9],
    [16, 0],    # now we're at position 16 in reference, past all the gaps
    [17, 0],
    [18, 0]],
```

## Immunological Epitopes and Experiments

The explorer dataset includes immunological epitopes from ViPR and IEDB that have been confirmed by experimentation. These data belong to the `explorer.models.Epitope` and `explorer.models.EpitopeExierment` models.

Epitopes have sequences that have been conformed to a particular alignment, so epitope sequences are defined by a protein MeSH ID and an alignment.

### Epitope Experiment Classes

#### Usage
`explorer/epitopeexperimentclasses?mesh_id=[MESH ID]&alignment=[ALIGNMENT NAME]`

**Example**: http://localhost:8000/explorer/epitopeexperimentclasses?mesh_id=D064370&alignment=20200505

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| host          | list | Distinct host organisms. |
| assay_type    | list | Distinct assay types. |
| assay_result  | list | Distinct assay result values. |
| mhc_allele    | list | Distinct MHC allele values. |
| mhc_class     | list | Dicting MHC class values. |
| exp_method    | list | Distinct experimental method values. |
| measurement_type | list | Distinct measurement type values. |
| iedb_id       | list | Distinct IEDB IDs. |

**Example**:
```
{
    "host": ["Human", "Mouse", "Rabbit", "Chicken", "Chimpanzee", "Guinea pig", "Monkey", "Humanized (transgene)"], "assay_type": ["B Cell", "MHC Binding", "T Cell"],
    "assay_result": ["Positive", "Negative", "Positive-Low", "Positive-High", "Positive-Intermediate"],
    "mhc_allele": ["", "HLA-A*33:01", "HLA-A*03:01", "HLA-A*11:01", "HLA-A*68:01", "HLA-A*31:01", "HLA-DRB1*01:01", "HLA-B*18:01", "HLA-B*40:01", "HLA-B*40:02", "HLA-B*44:03", "HLA-B*44:02", "HLA-B*45:01", "Patr-B*24:01", "HLA-A*23:01", "HLA-A*24:02", "HLA-A*26:01", "HLA-A*29:02", "HLA-A*01:01", "HLA-A*30:02", "HLA-A*02:01", "HLA-A*68:02", "HLA-A*02:02", "HLA-A*02:03", "HLA-A*02:06", "HLA-B*35:01", "HLA-B*51:01", "HLA-B*53:01", "HLA-B*07:02", "HLA-B*54:01", "HLA-DRB1*04:01", "H2-d class II", "Patr-A*01:01", "H2-Db", "H2-Kb", "H2-b class I", "Patr-B*01:01", "H2-d class I", "Patr-A*07:01", "Patr-A*09:01", "Mamu-B*001:01", "HLA-B*08:01", "HLA-B*15:01", "HLA-B*27:05", "HLA-B*58:01", "HLA-A*68:23", "HLA-A*32:15", "HLA-A*69:01", "HLA-B*15:42", "HLA-B*45:06", "HLA-C*04:01", "HLA-A2", "H2 class I", "H2-b class II", "Patr-B*13:01", "HLA-C*07:02", "Mamu-B*017:04", "HLA-DRA*01:01/DRB1*07:01", "H2-Kd", "HLA-C*15:02", "HLA class II", "HLA-DR8", "HLA-C*14:02", "H2-k class I", "H2-k class II", "H2-IEk", "H2-IAb"],
    "mhc_class": ["", "I", "II"],
    ...
}
```

### Epitope Experiments

#### Usage
`explorer/epitopeexperimentsfilter?alignment=[ALIGNMENT NAME]&mesh_id=[MESH ID]&[ADDITIONAL CRITERIA]`

**Example**: http://localhost:8000/explorer/epitopeexperimentsfilter?alignment=20200505&mesh_id=D064370&host=Human&mhc_class=I&assay_result=Positive-High&iedb_id=73883

#### Arguments
| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |
| host          | N | Comma-separated list of host organisms for the epitope experiment. |
| assay_type    | N | Comma-separated list of assay types. |
| assay_result  | N | Comma-separated list of categorical string values describing assay result. |
| mhc_allele    | N | Comma-separated list of MHC alleles. |
| mhc_class     | N | Comma-separated list of MHC classes. |
| exp_method    | N | Comma-separated list of experimental methods. |
| measurement_type | N | Comma-separated list of measurement types. |
| iedb_id       | N | Comma-separated list of IEDB IDs for epitopes (e.g. for all epitopes having the MeSH and alignment). |


#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| host          | str | Host organism for the epitope experiment. |
| assay_type    | str | Type of assay. |
| assay_result  | str | Categorical string value describing assay result. |
| mhc_allele    | str | MHC allele if specified. |
| mhc_class     | str | MHC class if specified. |
| exp_method    | str | Experimental method used. |
| measurement_type | str | Description of the measurement meaning. |
| iedb_id       | str | Filter experiments to those involving epitope with this IEDB ID. |


**Example**:

```
[
    {
        "host": "Human",
        "assay_type": "MHC Binding",
        "assay_result": "Positive-High",
        "mhc_allele": "HLA-A*23:01",
        "mhc_class": "I",
        "exp_method": "purified MHC/competitive/radioactivity",
        "measurement_type": "dissociation constant KD (~IC50)",
        "iedb_id": "73883"
    }, {
        "host": "Human",
        "assay_type": "MHC Binding",
        "assay_result": "Positive-High",
        "mhc_allele": "HLA-A*24:02",
        "mhc_class": "I",
        "exp_method": "purified MHC/competitive/radioactivity",
        "measurement_type": "dissociation constant KD (~IC50)",
        "iedb_id": "73883"
    }, {
        "host": "Human",
        "assay_type": "MHC Binding",
        "assay_result": "Positive-High",
        "mhc_allele": "HLA-A*29:02",
        "mhc_class": "I",
        "exp_method": "purified MHC/competitive/radioactivity",
        "measurement_type": "dissociation constant KD (~IC50)",
        "iedb_id": "73883"
    },
    ...
]
```

### Epitope Sequences

#### Usage
`explorer/epitopesequence?alignment=[ALIGNMENT NAME]&mesh_id=[MESH ID]&iedb_id=[LIST OF IEDB IDS]`

**Example**: http://localhost:8000/explorer/epitopesequence?alignment=20200505&mesh_id=D064370&iedb_id=73883%2C1220%2C2770

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |
| iedb_id   | Y | Comma-separated list of iedb_id's whose aligned sequences you want. |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| iedb_id | str | IEDB ID of the epitope sequence. |
| offset | int | Offset of the epitope sequence in the alignment. |
| seq | str | Epitope sequence alignment with leading gaps trimmed. |

**Example**:
```
[
    {
        "iedb_id": "2770",
        "offset": 470,
        "seq": "ALNCYW---------------------------PLNDY--------------------------GFY-----------------------------------T-----TTGIGYQPYRVVVLSFEL"
    }, {
        "iedb_id": "1220",
        "offset": 988,
        "seq": "AEVQIDRLI"
    }, {
        "iedb_id": "73883",
        "offset": 1055,
        "seq": "YFPREGVFVF"
    },
    ...
]
```

## Crystal Structures

Available crystal structures and the chains we have for them can be accessed by protein MeSH ID through `structurechains`. Aligned amino acid chain sequences can be accessed by PDB ID and chain ID through `structuresequence`. Atomic coordinate data for each residue in a chain can be accessed through `structureatoms`, which will return atomic data for a single atom type across all residues by PDB and chain ID.

### Crystal Structure Names and Chains

Use `structurechains` to get a list of all available PDB IDs and chains.

#### Usage
`explorer/structurechains?mesh_id=[MESH ID]`

**Example**: http://localhost:8000/explorer/structurechains?mesh_id=D064370

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| taxon | str | Name of the Taxon specified by the structure. |
| taxon_id | str | NCBI Taxon ID. |
| pdb_id | str | PDB ID for the structure. |
| chain | char | Chain in the structure. |

**Example**:
```
[
    {
        "taxon": "Severe acute respiratory syndrome-related coronavirus",
        "taxon_id": "694009",
        "pdb_id": "5WRG",
        "chain": "A"
    }, {
        "taxon": "Severe acute respiratory syndrome-related coronavirus",
        "taxon_id": "694009",
        "pdb_id": "5WRG",
        "chain": "B"
    }, {
        "taxon": "Severe acute respiratory syndrome-related coronavirus",
        "taxon_id": "694009",
        "pdb_id": "5WRG",
        "chain": "C"
    }, {
        "taxon": "Human coronavirus HKU1 (isolate N1)",
        "taxon_id": "443239",
        "pdb_id": "5GNB",
        "chain": "A"
    }, {
        "taxon": "Middle East respiratory syndrome-related coronavirus",
        "taxon_id": "1335626",
        "pdb_id": "5X5F",
        "chain": "A"
    },
    ...
]
```

### Crystal Structure Sequence Alignments

Use `structuresequence` to get one or more aligned amino acid chain sequences.

#### Usage
`explorer/structuresequence?mesh_id=[MESH ID]&alignment=[ALIGNMENT]&pdbchains=[PDB.CHAIN LIST]`

**Example**: http://localhost:8000/explorer/structuresequence?mesh_id=D064370&alignment=20200505&pdbchains=5X5B.A%2C5X5B.C

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence's alignment. |
| pdbchains | Y | comma-separated list of [PDB ID].[CHAIN]. |

#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| pdbchain | str | [PDB_ID].[CHAIN] |
| pdb_id | str | PDB ID. |
| chain | char | Chain ID. |
| sequence | str | Aligned animo acid sequence from chain, leading gaps trimmed. |
| offset | int | Offset of sequence in alignment. |

**Example**:

```
[
    {
        "pdbchain": "5X5B.A",
        "pdb_id": "5X5B",
        "chain": "A",
        "sequence": "RCTTFD---------DVQA-------PNYTQHTSSMRGVYYPDEIFRSDTLYLTQDLFLPFYSNVTGFH--TINH--------------------TFDNPVIPFKDGIYFAATEKSNV-------------------VRGWVFGSTMNNKSQSVII--IN--------------------NSTNVVIRACN---------FELCDNPFFAVSKP-M-GT---QTHTMIFDNAFNCTFEYISDAFSLDVSE-KSGNF--KHLREFVFKNKDGFLYVYKGYQPIDVV----RDLPSGFNTLKPIFKLPLGINITNFRAILTAFS--------------TWGTSAAAYFVGYLKPTTFMLKYDENGTITDAVDCSQNPLAELKCSVKSFEIDKGIYQTSNFRVVPSGDVVR--------LCPFGEVF--NATKFPSVYAWERKKISNCVADYSVLYNS-TFFSTFKCYGVSATKLNDLCFSNVYADSFVVKGDDVRQIAPGQTGVIADYNYKLPDDFMGCVLAWNTRNIDATS---------------------------TGNYNYKYRYLRHGKLRPFERDISNVPFS-------------------PDGKPCT-PPALNCYW---------------------------PLNDY--------------------------GFY-----------------------------------T-----TTGIGYQPYRVVVLSFE---------------------DLIKNQCVNFNFNGLTGTGVLTP-SSKRFQPFQQFGRDVSDFTDSVRDPKTSEILDISPCSFGGVSVITPGTNASSEVAVLYQDVNCTDVSTAIHADQLTPAWRIYST-----GNNVFQTQAGCLIGAEHVDT----SYECDIPIGAGICASYHT-------------------KSIVAYTMSLG-ADSSIAYSNNTIAIPTNFSISITTEVMPVSMAKTSVDCNMYICGDSTECANLLLQYGSFCTQLNRALSGIAAEQDRNTREVFAQVKQMY--KTPTLK----YFGGFNFSQILPDPL-----------KPTKRSFIEDLLFNKV-----------------------TDLICAQKFNGLTVLPPLLTDDMIAAYTAALVSGTATAGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKQIANQFNKAISQIQESLTTTSTALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQAAPHGVVFLHVTYVPSQERNFTTAPAICHE---GKAYFPREGVFVFNGT----------SWFITQRNFFSPQIIT-TDNTFV---------------------------------------------------------------------------------------------------------------------------------------------------------------",
        "offset": 35
    }, {
        "pdbchain": "5X5B.C",
        "pdb_id": "5X5B",
        "chain": "C",
        "sequence": "RCTTFD---------DVQA-------PNYTQHTSSMRGVYYPDEIFRSDTLYLTQDLFLPFYSNVTGFH--TINH--------------------TFDNPVIPFKDGIYFAATEKSNV-------------------VRGWVFGSTMNNKSQSVII--IN--------------------NSTNVVIRACN---------FELCDNPFFAVSKP-M-GT---QTHTMIFDNAFNCTFEYISDAFSLDVSE-KSGNF--KHLREFVFKNKDGFLYVYKGYQPIDVV----RDLPSGFNTLKPIFKLPLGINITNFRAILTAFS--------------TWGTSAAAYFVGYLKPTTFMLKYDENGTITDAVDCSQNPLAELKCSVKSFEIDKGIYQTSNFRVVPSGDVVRF-PNITN-LCPFGEVF--NATKFPSVYAWERKKISNCVADYSVLYNS-TFFSTFKCYGVSATKLNDLCFSNVYADSFVVKGDDVRQIAPGQTGVIADYNYKLPDDFMGCVLAWNTRNIDATS---------------------------TGNYNYKYRYLRHGKLRPFERDISNVPFS-------------------PDGKPCT-PPALNCYW---------------------------PLNDY--------------------------GFY-----------------------------------T-----TTGIGYQPYRVVVLSFELLNAPATVCGPK-L-----STDLIKNQCVNFNFNGLTGTGVLTP-SSKRFQPFQQFGRDVSDFTDSVRDPKTSEILDISPCSFGGVSVITPGTNASSEVAVLYQDVNCTDVSTAIHADQLTPAWRIYST-----GNNVFQTQAGCLIGAEHVDT----SYECDIPIGAGICASYHT-------------------KSIVAYTMSLG-ADSSIAYSNNTIAIPTNFSISITTEVMPVSMAKTSVDCNMYICGDSTECANLLLQYGSFCTQLNRALSGIAAEQDRNTREVFAQVKQMY--KTPTLK----YFGGFNFSQILPDPL-----------KPTKRSFIEDLLFNKV-----------------------T-LICAQKFNGLTVLPPLLTDDMIAAYTAALVSGTATAGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKQIANQFNKAISQIQESLTTTSTALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQAAPHGVVFLHVTYVPSQERNFTTAPAICHE---GKAYFPREGVFVFNGT----------SWFITQRNFFSPQIIT-TDNTFV---------------------------------------------------------------------------------------------------------------------------------------------------------------",
        "offset": 35
    },
    ...
]
```

### Crystal Structure Residue Info and Coordinates

Use `structureatoms` to get atomic data of a single atom type across all residues for specificed chains in PDB structures.

**Note:** Only one atom can be selected for all residues in a call. The example URL below gets all alpha carbons for two chains in two PDB structures. This will be changed in the future to provide useful information about side-chain interactions.

#### Usage
`explorer/structureresidueatoms?mesh_id=D064370&atom=CA&pdbchains=5X5B.A%2C5X5B.C`

**Example**: http://localhost:8000/explorer/structureresidueatoms?mesh_id=D064370&atom=CA&alignment=20200505&pdbchains=5X5B.A%2C5X5B.C

#### Arguments

| Argument | Required | Description |
|----------|----------|-------------|
| mesh_id   | Y | NCBI MeSH ID of the protein. |
| alignment | Y | Name of the sequence alignment to use. |
| pdbchains | Y | comma-separated list of [PDB ID].[CHAIN]. |
| atom      | Y | comma-separated list of atom names to retreive data for. |
#### Returns

An array of:

| Property | Type | Description |
|----------|------|-------------|
| pdb_id    | str   | PDB ID of the structure. |
| chain     | char  | Chain name. |
| pdb_chain  | str   | [PDB ID].[CHAIN], a unique identifier. |
| alignment | str | Name of the sequence alignment used. |
| atoms     | array | Array of atoms. |
| atoms[i].atom      | str   | Name of the atom. |
| atoms[i].element   | str   | Atomic symbol. |
| atoms[i].charge    | int   | Atomic charge. |
| atoms[i].occupancy | float | Atomic occupancy. |
| atoms[i].x         | float | X coordinate. |
| atoms[i].y         | float | y coordinate. |
| atoms[i].z         | float | z coordinate. |
| atoms[i].resn     | char   | Single-letter amino acid code for the residue that this atom is part of. |
| atoms[i].resid     | int   | Numeric identifier of the residue this atom belongs to, for the crystal structure file. |
| atoms[i].resix     | int   | Numeric index of the residue this atom belongs to in the raw, unaligned, ungapped amino acid sequence. |
| atoms[i].resaln    | int   | Numeric index of the residue this atom belongs to in the aligned amino acid sequence, using the alignment requested. |




**Example**:
```
[
    {
        "pdbchain": "5X5B.A",
        "pdb_id": "5X5B",
        "chain": "A",
        "alignment": "20200505",
        "residues": [
            {
                "resid": 18,
                "resix": 0,
                "resn": "R",
                "resaln": 30,
                "atom": "CA",
                "atom_x": 46.76,
                "atom_y": -31,
                "atom_z": 43,
                "element": "C",
                "charge": 0
            }, {
                "resid": 19,
                "resix": 1,
                "resn": "C",
                "resaln": 31,
                "atom": "CA",
                "atom_x": 49.18,
                "atom_y": -28,
                "atom_z": 42,
                "element": "C",
                "charge": 0
            },
            ...
    },
    ...
]
```




fin.
